File: benchmark/threads/threads_001.py
import timeit
from multiprocessing import Process, Queue

def process(r, ini, fim):
	acc = 0
	i = ini

	while (i <= fim):
		acc = acc + i
		i = i+1

	r.put(acc)


if __name__ == '__main__':

	t0 = timeit.default_timer()

	nthreads = 2
	L = 10000000
	n = nthreads * L
	acc = Queue()
	process(acc, 0, n)

	t1 = timeit.default_timer()

	

	t2 = timeit.default_timer()

	acc1 = Queue()
	acc2 = Queue()

	p1 = Process(target=process, args=(acc1, 0, L))
	p2 = Process(target=process, args=(acc2, L + 1, 2 * L))
	p1.start()
	p2.start()
	p1.join()
	p2.join()

	t3 = timeit.default_timer()

	print "Time elapsed multi-thread python: ", t3 - t2
	print "Time elapsed single-thread python: ", t1 - t0
	print "Relative time difference: ", 100 * (t1-t0 - t3+t2)/(t1-t0), "%" 
	print "Result: ", acc1.get() + acc2.get()

	

File: benchmark/threads/threads_001.rb
L = 10000000

tini = Time.now
def proc(ini, fim)
  acc = 0
  i = ini
  while i <= fim do
    acc += i
    i += 1
  end
  acc
end

th1 = Thread.new {
  Thread.current['v'] = proc(0, L)
}

th2 = Thread.new {
  Thread.current['v'] = proc(L+1, 2 * L)
}

th1.join
th2.join
tfim = Time.now

r1 = th1['v'] + th2['v'];
t1 = (tfim - tini) / 2

tini = Time.now
r2 = proc(0, 2 * L)
tfim = Time.now

t2 = tfim - tini

puts "Time elapsed single-thread version ruby: #{t2}"
puts "Time elapsed multi-thread version ruby: #{t1}"

if r1 == r2
  puts "Relative time difference: " + (100 * (t2-t1) / t2).to_s
	puts "Result: #{r1}"
else
  puts "Test threads_001.rb failed!"
end